accepted/single-file/design.md

```diff
@@ -0,0 +1,281 @@
+namespace System.Runtime.Loader
+{
+    public partial class Bundle
+    {
+        public static System.IO.Stream GetFileStream(Assembly assembly, string name);
+    }
+}
+```
+
+We can also provide an abstraction that abstracts away the physical location of a file (bundle or disk). For example, add a variant of  `GetFileStream` API that looks for a file in the bundle, and if not found, falls back to disk-lookup.
+
+#### Single-file Plugins
+
+The above design should be extended to seamlessly support single-file publish for plugins.
+
+- The bundler tool will mostly work as-is, regardless of whether we publish an application or class-lib. The binary blob with dependencies can be appended to both native and managed binaries.
+- For host/runtime support, the options are:
+  - Implement plugins using existing infrastructure. For example: Take control of assembly/native binary loads via existing `AssemblyLoadContext` callbacks and events. Extract the files embedded within the single-file plugin using the `GetFileStream()` API and load them on demand.
+  - Have new API to load a single-file plugin, for example: `AssemblyLoadContext.LoadWithEmbeddedDependencies()`.
+
+#### Project File Options
+
+By default, `PublishSingleFile` bundles all files in the publish directory into the single executable. This behavior can be altered via the following configurations in the application's project file.
+
+* Exclude certain files in the publish directory from being bundled (ex: because of licensing issues or for testing purposes).
+
+```xml
+<ItemGroup>
+    <SingleFileExclude Include="List of patterns to exclude"/>
+</ItemGroup>
+```
+
+- Add additional files to the bundle from other locations (ex: add data files to the app). The developer will be responsible for handling these additional files.
+
+```xml
+<ItemGroup>
+    <SingleFileAdd Include="List of file paths"/>
+</ItemGroup>
+```
+
+An app may choose to have all of its dependencies extracted to disk at runtime (instead of loading certain files directly from the bundle) by setting the following property.
+
+```xml
+<PropertyGroup>
+    <SingleFileExtractAll>true</SingleFileExtractAll>
+</PropertyGroup>
+```
+
+#### VS Integration
+
+Developers should be able to use the feature easily from Visual Studio. The feature will start with text based support -- by explicitly setting the `PublishSingleFile` property in the project file. In future, we may provide a single-file publish-profile, or other UI triggers.
+
+## User Experience
+
+To summarize, here's the overall experience for creating a HelloWorld single-file app 
+
+*  Create a new HelloWorld app: `HelloWorld$ dotnet new console`
+
+#### Framework Dependent HelloWorld
+
+* Normal publish: `dotnet publish` 
+
+  * Publish directory contains the host `HelloWorld.exe` ,  the app `HelloWorld.dll`, configuration files `HelloWorld.deps.json`, `HelloWorld.runtimeconfig.json`, and the PDB-file `HelloWorld.pdb`.
```

Comment:
Don’t framework-dependent publishes only include the app’s `.dll` and to run it, you have to run `dotnet HelloWorld.dll`? Where does this `.exe` come from?

Reply:
That was changed for .NET Core 3.0. You get `.exe` for framework dependent publishes by default in .NET Core 3.0.

Reply:
Oh cool, didn’t know that! So the `.exe` will then pick up the “correct” runtime if one is installed automatically? Do you happen to have a link for me where this process is described?

Reply:
You can find most of this process described [here](https://github.com/dotnet/core-setup/blob/master/Documentation/design-docs/host-components.md)